### PR TITLE
Fix svd_p to stop returning garbage values of u/vt when compute_uv ==…

### PR DESCRIPTION
**Pull request by bchetioui**
… False.

The custom call made by svd_p does not compute u and vt when
compute_uv is set to False. Returning them using the primitive
means that it is up to the caller to throw away these values.

### CODE REVIEW

File: jax/lax_linalg.py [hunk 1]
```diff
@@ -847,9 +854,9 @@ def _qr_cpu_gpu_translation_rule(geqrf_impl, orgqr_impl, c, operand,
 # Singular value decomposition
 
 def svd_impl(operand, full_matrices, compute_uv):
-  s, u, vt = xla.apply_primitive(svd_p, operand, full_matrices=full_matrices,
-                                 compute_uv=compute_uv)
-  return s, u, vt
+  result = xla.apply_primitive(svd_p, operand, full_matrices=full_matrices,
```
Comment (hawkinsp) [inline jax/lax_linalg.py:857]: (In the past, unpacking this sequence was actually necessary, but I don't think it is any more.)

Can we just `return xla.apply_primitive ...`

File: jax/lax_linalg.py [hunk 2]
```diff
@@ -872,11 +884,14 @@ def svd_abstract_eval(operand, full_matrices, compute_uv):
...
+      u = ShapedArray(batch_dims + (m, m if full_matrices else min(m, n)), operand.dtype)
+      vt = ShapedArray(batch_dims + (n if full_matrices else min(m, n), n), operand.dtype)
+      return s, u, vt
+    else:
+      return s,
   else:
     raise NotImplementedError
-  return s, u, vt
 
 def svd_jvp_rule(primals, tangents, full_matrices, compute_uv):
```
Comment (hawkinsp) [inline jax/lax_linalg.py:896]: I think you need to fix the JVP rule also? The output signature of the JVP rule should mirror that of the regular primitive.

**Review by hawkinsp — CHANGES_REQUESTED** [review]

File: jax/lax_linalg.py (hunk 2 above)
Comment (SIben) [reply jax/lax_linalg.py:895]: Addressed in last commit.
Yes, there is a in `lax/linalg_test.py`, which somehow is not failing. I am not entirely sure why yet?

Comment (hawkinsp) [reply jax/lax_linalg.py:895]: One thing that is always worth checking is that we aren't failing to sample such a test. We only run a random sample of possible tests. Try setting `JAX_NUM_GENERATED_CASES=100` (or larger) and running that test case.

File: jax/lax_linalg.py [hunk 3]
```diff
@@ -900,6 +914,9 @@ def svd_jvp_rule(primals, tangents, full_matrices, compute_uv):
   dU = jnp.matmul(U, F * (dSS + _T(dSS)))
   dV = jnp.matmul(V, F * (SdS + _T(SdS)))
 
+  if not compute_uv:
```
Comment (gnecula) [inline jax/lax_linalg.py:917]: Since you are returning early anyway, why not return as soon as you have `ds`?

**Review by gnecula — APPROVED** [review]
LGTM

File: jax/lax_linalg.py (hunk 3 above)
Comment (SIben) [reply jax/lax_linalg.py]: Addressed in last commit!

File: jax/lax_linalg.py (hunk 2 above)
Comment (SIben) [reply jax/lax_linalg.py:895]: The branch is definitely hit, but there is no reported failure with or without the fix to the jvp rule, it is odd.

Comment (SIben) [reply jax/lax_linalg.py:896]: This had to do with the way I unpacked the value in `svd`. I fixed it in the last commit, this is resolved.

**Pull request merged**